pgx/src/heap_tuple.rs

```diff
@@ -0,0 +1,340 @@
+            Some((attnum, _)) => self.set_by_index(attnum, value),
+        }
+    }
+
+    /// Given the index for an attribute in this [PgHeapTuple], change its value.
+    ///
+    /// Attribute numbers start at 1, not 0.
+    ///
+    /// ## Errors
+    /// - return [TryFromDatumError::NoSuchAttributeNumber] if the attribute does not exist
+    /// - return [TryFromDatumError::IncompatibleTypes] if the Rust type of the `value` is not
+    /// compatible with the attribute's Postgres type
+    pub fn set_by_index<T: IntoDatum>(
+        &mut self,
+        attno: NonZeroUsize,
+        value: T,
+    ) -> FromDatumResult<()> {
+        unsafe {
+            match self.get_attribute_by_index(attno) {
+                None => return Err(TryFromDatumError::NoSuchAttributeNumber(attno)),
+                Some(att) => {
+                    if !T::is_compatible_with(att.atttypid) {
+                        return Err(TryFromDatumError::IncompatibleTypes);
+                    }
+                }
+            }
+
+            let mut datums = (0..self.tupdesc.len()).collect::<Vec<_>>();
+            let mut nulls = (0..self.tupdesc.len()).map(|_| false).collect::<Vec<_>>();
+            let mut do_replace = (0..self.tupdesc.len()).map(|_| false).collect::<Vec<_>>();
+
+            let datum = value.into_datum();
+            let attno = attno.get() - 1;
+
+            nulls[attno] = datum.is_none();
+            datums[attno] = datum.unwrap_or(0);
+            do_replace[attno] = true;
+
+            let new_tuple = PgBox::<pg_sys::HeapTupleData, AllocatedByRust>::from_rust(
+                pg_sys::heap_modify_tuple(
+                    self.tuple.as_ptr(),
+                    self.tupdesc.as_ptr(),
+                    datums.as_mut_ptr(),
+                    nulls.as_mut_ptr(),
+                    do_replace.as_mut_ptr(),
+                ),
+            );
+            let old_tuple = std::mem::replace(&mut self.tuple, new_tuple);
+            drop(old_tuple);
+            Ok(Some(()))
+        }
+    }
+}
+
+impl<'a, AllocatedBy: WhoAllocated<pg_sys::HeapTupleData>> PgHeapTuple<'a, AllocatedBy> {
+    /// Consume this [PgHeapTuple] and return a Datum representation, which is a pointer to the
+    /// underlying [pg_sys::HeapTupleData] struct.
+    pub fn into_datum(self) -> Option<pg_sys::Datum> {
+        self.tuple.into_datum()
+    }
+
+    /// Consume this [PgHeapTuple] and return a Datum representation, which is a pointer to a
+    /// [pg_sys::HeapTupleHeaderData] struct, containing the tuple data and the corresponding
+    /// tuple descriptor information.
+    pub fn into_composite_datum(self) -> Option<pg_sys::Datum> {
+        unsafe {
+            Some(pg_sys::heap_copy_tuple_as_datum(
+                self.tuple.as_ptr(),
+                self.tupdesc.as_ptr(),
+            ))
+        }
+    }
+
+    /// How many attributes does this [PgHeapTuple] have?
```

Comment:
Generally doc comments that are questions are discouraged in favor of the more bland style. Copying `Vec`'s:

```rust
/// Returns the number of attributes in the [PgHeapTuple], also referred to as its ‘length’.
```